<a href="https://colab.research.google.com/github/miguelrendon107/Analitica/blob/main/Reto_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**RETO 1**

**Cargar las Librerias y base de datos**

In [ ]:
import numpy as np
import pandas as pd
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import confusion_matrix
from sklearn.metrics import accuracy_score

base= "/content/1. BD1_Longitudinal Survey.xlsx"
bd= pd.read_excel(base, sheet_name= 0)
bd= bd.dropna()

bd = bd[['Age', 'Father_Edu', 'Mother_Edu', 'Height', 'Weight', 'Income', 'Self_Esteem', 'FamilySize', 'Urban']]


bdx = np.array(bd[['Age', 'Father_Edu', 'Mother_Edu', 'Height', 'Weight', 'Income', 'Self_Esteem', 'FamilySize']])
bdy = np.array(bd['Urban'])




### **ANALISIS BASE DE DATOS**






***1. Cual es el numero total de datos de la base de datos?***

In [ ]:
t= len(bd) #Tamaño de la base de datos
print("Tamaño de la base de datos:", t)

Tamaño de la base de datos: 5683


El tamaño de la base de datos es de 5683.

***2. Cual es el porcentaje de los datos que hacen parte de cada una de las categorias de la selección?***



In [ ]:
#Personas que viven en la ciudad.
filas=np.where(bdy==1)[0]
len(filas)
print("Porcentaje de las personas que viven en la ciudad:", len(filas)/len(bd)*100,"%")

#Personas que viven en areas rurales.
filas=np.where(bdy==0)[0]
len(filas)
print("Porcentaje de las personas que viven en las areas rurales:", len(filas)/len(bd)*100,"%")

Porcentaje de las personas que viven en la ciudad: 79.35949322540911 %
Porcentaje de las personas que viven en las areas rurales: 20.640506774590886 %


- Porcentaje de las personas que viven en la ciudad: 79.35949322540911 %

- Porcentaje de las personas que viven en las areas rurales: 20.640506774590886 %

# **MODELO**

In [ ]:
#Modelo de Naive Bayes
mnb= GaussianNB()
#Entrenamiento del modelo
mnb.fit(bdx, bdy)
#Predicción de los datos de entrenamiento
ydp= mnb.predict(bdx)

#Digame las medias de las variables para cada categoria -
u=mnb.theta_ #se estiman las medias
sigma=np.sqrt(mnb.var_) #se estiman las varianzas
print("Las Medias son:", u)
print("Las Varianzas son:", sigma)

Las Medias son: [[1.83810742e+01 1.03324808e+01 1.07612958e+01 6.72148338e+01
  1.48857630e+02 3.03672506e+04 2.20272805e+01 4.74168798e+00]
 [1.83569845e+01 1.12565410e+01 1.11769401e+01 6.69031042e+01
  1.44760976e+02 3.34278929e+04 2.26004435e+01 4.68780488e+00]]
Las Varianzas son: [[2.48138660e+00 3.86936748e+00 3.17237004e+00 4.07785391e+00
  3.01106662e+01 3.17019089e+04 4.19281202e+00 2.57546964e+00]
 [2.47587326e+00 4.10556364e+00 3.38680150e+00 4.21331089e+00
  3.00913213e+01 3.61232282e+04 4.18499090e+00 2.43784107e+00]]


In [ ]:
#Las variaciones de las variables
#'Age', 'Father_Edu', 'Mother_Edu', 'Height', 'Weight', 'Income', 'Self_Esteem', 'FamilySize'
#Los resultados de la primera matriz son para 0 (areas rurales) y la segunda matriz es para los 1 (areas urbanas)
LI= u-sigma #Limite inferior
LS= u+sigma #Limite superior
print("Los limites inferiores son:", LI)
print("Los limites superiores son:", LS)

Los limites inferiores son: [[ 1.58996876e+01  6.46311334e+00  7.58892579e+00  6.31369799e+01
   1.18746964e+02 -1.33465823e+03  1.78344685e+01  2.16621834e+00]
 [ 1.58811112e+01  7.15097738e+00  7.79013864e+00  6.26897933e+01
   1.14669654e+02 -2.69533531e+03  1.84154526e+01  2.24996381e+00]]
Los limites superiores son: [[2.08624608e+01 1.42018483e+01 1.39336659e+01 7.12926877e+01
  1.78968296e+02 6.20691595e+04 2.62200925e+01 7.31715762e+00]
 [2.08328577e+01 1.53621047e+01 1.45637416e+01 7.11164151e+01
  1.74852297e+02 6.95511211e+04 2.67854344e+01 7.12564594e+00]]


***3. Intervalos variables:***
* **Edad** -> rural: [15-20]; urbana [15-20]
* **Educación padre** -> rural: [6-14]; urbana [7-15]
* **Educacion madre** -> rural: [7-13]; urbana [7-14]
* **Altura** -> rural: [63.1-71.2]; urbana: [62.2-71.1]
* **Peso** -> rural: [118.7-178.9]; urbana: [114.6-174.8]
* **Ingresos** -> rural: [-1,334.6-6,206.9]; urbana: [-2,695.3-6,955.1]
* **Autoestima** -> rural: [17.8-26.26.2]; urbana: [18.4-26.7]
* **Tamaño Familia** -> rural: [2-7]; urbana: [2-7]

In [ ]:
# Calcular la matriz de confusión
cm= confusion_matrix(bdy, ydp)
print("Matriz de confusión:", cm)

Matriz de confusión: [[   2 1171]
 [   4 4506]]


In [ ]:
VN=cm[0,0] #Verdaderos negativos
FP=cm[0,1] #Falsos positivos
FN=cm[1,0] #Falsos negativos
VP=cm[1,1] #Verdaderos positivos


print("Exactitud:",  (VP+VN)/(VP+VN+FP+FN)) #Comportamiento general
print("Tasa Error:", (FP+FN)/(VP+VN+FP+FN)) #Porcentaje equivocaciones
print("Sensibilidad:", VP/(VP+FP)) #Como se comporto frente a los positivos
print("Especificidad:", VN/(VN+FN)) #Como se comporto frente a los negativos

print("Precision:", VP/(VP+FN)) #Como se comporta frente a los positivos
print("La probabilidad de negativos:", VN/(VN+FP)) #Como se comporta frente a los clasificados negativos (areas rurales)

Exactitud: 0.7932430054548654
Tasa Error: 0.2067569945451346
Sensibilidad: 0.793729082261758
Especificidad: 0.3333333333333333
Precision: 0.9991130820399113
La probabilidad de negativos: 0.0017050298380221654


***4. Interpretación metricas:***

- **Exactitud:** El modelo clasifica correctamente el **79.32%** de las instancias totales.
- **Tasa Error:** El modelo clasifica incorrectamente el 20.68% de las instancias totales.
- **Sensibilidad:** De todas las instancias que son realmente positivas (ciudad), el modelo identifica correctamente el 79.37%
-  **Especificidad:** De todas las instancias que son realmente negativas (rural), el modelo identifica correctamente el 33.33%
- **Precisión:** De todas las instancias que el modelo predice como positivas (ciudad), el 99.91% son realmente positivas.
- **Valor predicción negativos:** De todas las instancias que el modelo predice como negativas (rural), el 0.17% son realmente positivas.

***5. Segun las caracteristicas brindadas en el excel:***

In [ ]:
Xun= [[34, 9, 6, 65, 135, 50000, 30, 3]] #Esta es la persona del Excel.
ydun=mnb.predict(Xun)

#Si el resultado es 1 la persona vive en la ciudad pero si es 0 vive en areas rurales.
if ydun[0] == 1:
    print("La persona vive en la ciudad.")
else:
    print("La persona vive en áreas rurales.")

La persona vive en la ciudad.


# **ANALISIS DE RESULTADOS:**

Despues de analizar los datos y aplicar un modelo Naive Bayes, los resultados sugieren en primer lugar que los intervalos de las variables de edad y tamaño de la familia son similares entre las personas que viven en áreas rurales y urbanas, existen diferencias notables en la educación, ingresos y autoestima. Los padres y madres en áreas urbanas tienden a tener un nivel de educación más alto, y las personas en áreas urbanas muestran una mayor autoestima y un rango de ingresos más amplio.

Por otro lado, el modelo clasifica correctamente el 79.32% de las instancias totales. Asi mismo, los resultados indican una alta precisión (99.91%) y sensibilidad (79.37%) al identificar a las personas que viven en la ciudad, lo que significa que el modelo es muy eficaz en este aspecto. Sin embargo, la especificidad (33.33%) y el valor de predicción negativa (0.17%) son bajos, lo que sugiere que el modelo es menos eficaz para identificar correctamente a las personas que viven en áreas rurales.

En conclusión , el modelo muestra una muy buena capacidad de clasificación de las personas que viven en areas urbanas, pero tiene una capacidad limitada para clasificar correctamente a los que viven en areas rurales.

**NOTA:** Estas diferencias destacan la necesidad de ajustar el modelo para mejorar su capacidad predictiva en áreas rurales y así obtener una representación más precisa de ambas poblaciones.


# EXTRA

In [ ]:
# Ayuda a entender cómo se distribuyen las diferentes categorías dentro de cada variable.
p1 = bd['Age'].value_counts(normalize=True) * 100
print("Porcentaje de datos por categoría:", p1)
p2 = bd['Father_Edu'].value_counts(normalize=True) * 100
print("Porcentaje de datos por categoría:", p2)
p3 = bd['Mother_Edu'].value_counts(normalize=True) * 100
print("Porcentaje de datos por categoría:", p3)
p4 = bd['Height'].value_counts(normalize=True) * 100
print("Porcentaje de datos por categoría:", p4)
p5 = bd['Weight'].value_counts(normalize=True) * 100
print("Porcentaje de datos por categoría:", p5)
p6 = bd['Income'].value_counts(normalize=True) * 100
print("Porcentaje de datos por categoría:", p6)
p7= bd['Self_Esteem'].value_counts(normalize=True) * 100
print("Porcentaje de datos por categoría:", p7)
p8 = bd['FamilySize'].value_counts(normalize=True) * 100
print("Porcentaje de datos por categoría:", p8)

Porcentaje de datos por categoría: Age
17    14.235439
18    13.936301
19    13.672356
16    13.584374
15    11.543199
20    11.261658
21    11.032905
22    10.733767
Name: proportion, dtype: float64
Porcentaje de datos por categoría: Father_Edu
12.0    34.594404
16.0     8.534225
8.0      7.636812
10.0     6.686609
14.0     5.507654
11.0     5.454865
9.0      5.226113
6.0      4.451874
13.0     2.850607
7.0      2.709836
18.0     2.164350
4.0      1.918001
0.0      1.812423
3.0      1.794827
20.0     1.777230
5.0      1.742038
15.0     1.689249
17.0     1.530882
2.0      1.002991
19.0     0.527890
1.0      0.387119
Name: proportion, dtype: float64
Porcentaje de datos por categoría: Mother_Edu
12.0    42.688721
11.0     8.710188
10.0     7.865564
16.0     6.405068
8.0      5.666021
14.0     5.190920
9.0      4.821397
13.0     3.519268
6.0      3.360901
7.0      1.794827
15.0     1.513285
5.0      1.513285
4.0      1.425304
3.0      1.372515
17.0     1.073377
0.0      1.073377
18.0     